gpytorch/lazy/lazy_tensor.py

```diff
@@ -732,6 +734,217 @@ def add_jitter(self, jitter_val=1e-3):
+        new_root[..., :m, :n] = L
+        new_root[..., m:, : lower_left.shape[-1]] = lower_left
+        new_root[..., m:, n : (n + schur_root.shape[-1])] = schur_root
+
+        # Use pseudo-inverse of Z as new inv root
+        if l_is_triang:
+            # we know L is triangular, so inverting is a triangular solve agaist the identity
+            # we don't need the batch shape here, thanks to broadcasting
+            Eye = torch.eye(new_root.shape[-2], device=new_root.device, dtype=new_root.dtype)
+            new_inv_root = torch.triangular_solve(Eye, new_root, upper=False)[0]
+            new_inv_root = new_inv_root.transpose(-1, -2)
+        else:
+            Q, R = stable_qr(new_root)
+
+            if R.shape[-2] == R.shape[-1]:
+                new_inv_root = torch.triangular_solve(Q.transpose(-2, -1), R, upper=True).solution.transpose(-2, -1)
+            else:
+                # solve with least squares
+                new_inv_root = torch.lstsq(Q.transpose(-2, -1), R).solution.transpose(-2, -1)
+
+        add_to_cache(new_lazy_tensor, "root_decomposition", RootLazyTensor(TriangularLazyTensor(new_root)))
+        add_to_cache(new_lazy_tensor, "root_inv_decomposition", RootLazyTensor(TriangularLazyTensor(new_inv_root)))
+
+        return new_lazy_tensor
+
+    def add_low_rank(
+        self,
+        low_rank_mat,
+        root_decomp_method: Optional[str] = None,
+        root_inv_decomp_method: Optional[str] = None,
+        generate_roots: Optional[bool] = True,
+    ):
+        """
+        Adds a low rank matrix to the matrix that this LazyTensor represents, e.g.
+        computes A + BB^T. We then update both the tensor and its root decomposition.
+
+        We have access to, L and M where A \approx LL^T and A^{-1} \approx MM^T.
+        We compute \tilde{A} = A + BB^T = L(I + M B B^T M^T)L' and then decompose
+        (I + M VV^T M^T) \approx RR^T, using LR as our new root decomposition.
+
+        This strategy is described in more detail in "Kernel Interpolation for Scalable Online Gaussian
+        Processes," Stanton et al, AISTATS, 2021. https://arxiv.org/abs/2103.01454.
+
+        Args:
+            low_rank_mat (:obj:`torch.tensor`): the matrix `B` that we are adding to `A`.
+            root_decomp_method (:obj:`str`): how to compute the root decomposition of `A`.
+            root_inv_decomp_method (:obj:`str`): how to compute the root inverse decomposition of `A`.
+            generate_roots (:obj:`bool`): whether to generate the root decomposition of :math:`A` even if it
+            has not been created yet.
+
+        Returns:
+            :obj:`SumLazyTensor`: addition of A and BB^T.
+        """
+        from . import lazify
+        from .root_lazy_tensor import RootLazyTensor
+        from .sum_lazy_tensor import SumLazyTensor
+        from .triangular_lazy_tensor import TriangularLazyTensor
+
+        if not isinstance(self, SumLazyTensor):
```

Comment:
I don't think this if block is necessary - this casing happens as part of the SLT constructor.

Reply:
I don't see it in [here](https://github.com/cornellius-gp/gpytorch/blob/d433003402e09195a76054b0fb7c3cf4aa9189dd/gpytorch/lazy/sum_lazy_tensor.py#L14). This is really designed to prevent caching memory issues where we continually are adding low rank matrices and then the recursion just builds up a lot (have experienced it in previous iterations of this code). E.g. the situation like `A_T = A_0 + \sum_{i=1}^T a_i a_i' = A_{T-1} + a_T a_T`